File: 030_DeleteTheMiddleNodeOfALinkedList.cpp
// Topic: Linked List / Two pointers
// Delete the Middle Node of a Linked List: Find and remove the middle node from a linked list.
// Time Complexity: O(n)
// Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* deleteMiddle(ListNode* head)
{
    if (!head || !head->next)
        return nullptr;
    ListNode* slow = head;
    ListNode* fast = head->next->next;
    while (fast && fast->next)
    {
        slow = slow->next;
        fast = fast->next->next;
    }
    // slow is node before middle
    ListNode* todel = slow->next;
    slow->next = slow->next->next;
    delete todel;
    return head;
}

void printList(ListNode* h)
{
    while (h)
    {
        cout << h->val;
        if (h->next)
            cout << "->";
        h = h->next;
    }
    cout << '\n';
}

int main()
{
    ListNode* a = new ListNode(1);
    a->next = new ListNode(3);
    a->next->next = new ListNode(4);
    a->next->next->next = new ListNode(7);
    a->next->next->next->next = new ListNode(1);
    a->next->next->next->next->next = new ListNode(2);
    a->next->next->next->next->next->next = new ListNode(6);
    cout << "orig: ";
    printList(a);
    auto r = deleteMiddle(a);
    cout << "after: ";
    printList(r);

    ListNode* b = new ListNode(1);
    cout << "single orig: ";
    printList(b);
    auto r2 = deleteMiddle(b);
    cout << "single after: ";
    if (!r2)
        cout << "(empty)\n";
    else
        printList(r2);
    return 0;
}

File: 030_DeleteTheMiddleNodeOfALinkedList.py
"""
Topic: Linked List / Two Pointers
Delete The Middle Node Of A Linked List: remove middle using slow/fast pointers
Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def delete_middle(head: Optional[ListNode]) -> Optional[ListNode]:
    if not head or not head.next:
        return None
    slow = head; fast = head; prev = None
    while fast and fast.next:
        prev = slow
        slow = slow.next
        fast = fast.next.next
    # slow is middle, prev is before it
    prev.next = slow.next
    return head

def make_list(vals):
    dummy = ListNode(0); cur = dummy
    for v in vals: cur.next = ListNode(v); cur = cur.next
    return dummy.next

def to_list(head):
    res=[]
    while head: res.append(head.val); head = head.next
    return res

if __name__ == '__main__':
    h = make_list([1,3,4,7,1,2,6])
    print(to_list(delete_middle(h)))  # [1,3,4,1,2,6]
    h2 = make_list([1,2,3,4])
    print(to_list(delete_middle(h2)))  # [1,2,4]
